src/models/request/tokenRequest.ts

```diff
@@ -22,13 +31,21 @@ export class TokenRequest {
 
     toIdentityToken(clientId: string) {
         const obj: any = {
-            grant_type: 'password',
-            username: this.email,
-            password: this.masterPasswordHash,
             scope: 'api offline_access',
             client_id: clientId,
         };
 
+        if (this.masterPasswordHash != null && this.email != null) {
+            obj.grant_type = 'password';
+            obj.username = this.email;
+            obj.password = this.masterPasswordHash;
+        } else if (this.code != null && this.codeVerifier != null && this.redirectUri != null) {
+            obj.grant_type = 'authorization_code';
+            obj.code = this.code;
+            obj.code_verifier = this.codeVerifier;
+            obj.redirect_uri = this.redirectUri;
+        }
```

Comment:
Is there a handler or possibility that this could ever have one final fall-out where `grant_type` is not set or validation around that? Say if `username` or `masterPasswordHash` are `null` and/or one of the `sso` parameters?